# PIX and Brazil Payments Trends – Exploratory Analysis

This notebook performs initial exploration and cleaning of the Brazilian Payment Methods dataset (2016–2024).  
The goal is to prepare the data for SQL aggregation and Tableau visualization, using pandas and numpy, focusing on PIX and the decline of traditional payment methods (DOC, TED, Boleto, Checks).

## Import Libraries

In [1]:
import pandas as pd
import numpy as np

## Load Dataset

In [2]:
df = pd.read_csv('../data/raw/brazilian_payment_methods_raw.csv') # Load Dataset

df.head(20) # Quick look at the data

YearMonth  quantityPix    valuePix  quantityTED    valueTED  quantityTEC  \
0      202405   5228826.97  2137205.79     69023.11  3482036.97         0.00   
1      202404   4924681.00  2083746.68     70647.15  3652606.48         0.00   
2      202403   4903392.34  1936188.21     67565.05  3333054.27         0.00   
3      202402   4367133.78  1718217.65     65805.87  2981226.45        60.01   
4      202401   4371541.99  1818473.97     70343.25  3337344.85       387.18   
5      202312   4804459.10  1933836.67     80202.26  3629758.88      1152.37   
6      202311   4231283.46  1740874.02     74398.37  3135994.81       754.56   
7      202310   4069683.33  1666062.39     71870.52  3325168.12       664.92   
8      202309   3846749.03  1539042.27     69407.76  3190049.86       483.31   
9      202308   3767327.85  1529362.91     76572.58  3406864.68       614.09   
10     202307   3532968.05  1422663.64     71737.31  3363337.34       667.68   
11     202306   3296756.82  1365737.17     73637.38  3658988.12       710.66   
12     202305   3189216.58  1343030.27     75894.99  3763925.98       620.33   
13     202304   2910507.92  1191585.30     69286.74  3234549.03       661.70   
14     202303   3001956.41  1280326.31     80528.82  3635596.42       680.54   
15     202302   2539124.77  1055639.24     70232.83  2889422.93       645.05   
16     202301   2575314.32  1111232.73     77982.34  3394583.49       632.37   
17     202212   2876769.59  1221473.50     91361.54  3810324.55      1272.38   
18     202211   2499996.99  1075433.80     83094.38  3323697.77       602.37   
19     202210   2436652.95  1046436.42     79833.93  3248099.31       657.19   

    valueTEC  quantityBankCheck  valueBankCheck  \
0       0.00           10467.85        40133.55   
1       0.00           11179.22        43295.40   
2       0.00           10122.46        36644.55   
3     109.32           10373.38        37076.26   
4     815.94           11919.12        41615.89   
5    2498.31           11241.91        39208.22   
6    1711.45           11866.27        40735.16   
7    1578.12           12902.42        46646.90   
8    1085.19           11916.38        41653.43   
9    1481.22           13113.98        46973.84   
10   1589.98           12782.70        43634.75   
11   1683.40           12898.39        44354.61   
12   1492.52           14171.91        50982.54   
13   1594.82           12188.20        43399.52   
14   1584.04           13763.41        55997.48   
15   1512.53           12393.22        50947.11   
16   1438.59           14091.41        46687.43   
17   2757.91           14246.95        47529.19   
18   1316.40           14501.40        48575.49   
19   1535.22           14985.60        49550.28   

    quantityBrazilianBoletoPayment  valueBrazilianBoletoPayment  quantityDOC  \
0                        349180.92                    506302.82         0.00   
1                        361633.28                    524619.69         0.00   
2                        336435.46                    469047.59         0.00   
3                        332269.47                    463066.01        61.89   
4                        363550.32                    516702.44       600.76   
5                        337552.05                    484391.91      1412.03   
6                        345234.36                    490503.75      1506.38   
7                        362721.59                    512336.66      1746.38   
8                        340115.13                    471176.87      2636.88   
9                        360865.30                    501254.74      5253.95   
10                       353354.67                    479276.62      4722.99   
11                       348653.58                    475375.14      3810.80   
12                       366029.09                    509111.41      4139.19   
13                       326293.01                    443418.56      2378.17   
14                       362664.11                 

## Dataset Overview
Check column types, missing values, and basic statistics

In [3]:
df.info()
df.describe()
df.isna().sum() # Check for missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 13 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   YearMonth                       101 non-null    int64  
 1   quantityPix                     101 non-null    float64
 2   valuePix                        101 non-null    float64
 3   quantityTED                     101 non-null    float64
 4   valueTED                        101 non-null    float64
 5   quantityTEC                     101 non-null    float64
 6   valueTEC                        101 non-null    float64
 7   quantityBankCheck               101 non-null    float64
 8   valueBankCheck                  101 non-null    float64
 9   quantityBrazilianBoletoPayment  101 non-null    float64
 10  valueBrazilianBoletoPayment     101 non-null    float64
 11  quantityDOC                     101 non-null    float64
 12  valueDOC                        101 

YearMonth                         0
quantityPix                       0
valuePix                          0
quantityTED                       0
valueTED                          0
quantityTEC                       0
valueTEC                          0
quantityBankCheck                 0
valueBankCheck                    0
quantityBrazilianBoletoPayment    0
valueBrazilianBoletoPayment       0
quantityDOC                       0
valueDOC                          0
dtype: int64

> **Note:** 
Each number (0) tells us how many missing values there are in that column. 
0 means no missing data in that column — all rows have valid values. 
The dataset is complete for all columns.

## Convert YearMonth to datetime

In [4]:
df['YearMonth'] = pd.to_datetime(df['YearMonth'], format='%Y%m') # Convert YearMonth to datetime
df['year'] = df['YearMonth'].dt.year # Extract year for trend analysis
df['month'] = df['YearMonth'].dt.month  # Extract month for seasonal analysis

df.head()

YearMonth  quantityPix    valuePix  quantityTED    valueTED  quantityTEC  \
0 2024-05-01   5228826.97  2137205.79     69023.11  3482036.97         0.00   
1 2024-04-01   4924681.00  2083746.68     70647.15  3652606.48         0.00   
2 2024-03-01   4903392.34  1936188.21     67565.05  3333054.27         0.00   
3 2024-02-01   4367133.78  1718217.65     65805.87  2981226.45        60.01   
4 2024-01-01   4371541.99  1818473.97     70343.25  3337344.85       387.18   

   valueTEC  quantityBankCheck  valueBankCheck  \
0      0.00           10467.85        40133.55   
1      0.00           11179.22        43295.40   
2      0.00           10122.46        36644.55   
3    109.32           10373.38        37076.26   
4    815.94           11919.12        41615.89   

   quantityBrazilianBoletoPayment  valueBrazilianBoletoPayment  quantityDOC  \
0                       349180.92                    506302.82         0.00   
1                       361633.28                    524619.69         0.00   
2                       336435.46                    469047.59         0.00   
3                       332269.47                    463066.01        61.89   
4                       363550.32                    516702.44       600.76   

   valueDOC  year  month  
0      0.00  2024      5  
1      0.00  2024      4  
2      0.00  2024      3  
3     34.21  2024      2  
4    616.58  2024      1

The date was stored as **202405** and now it's **2024-05-01**.

> **Why?**  
> Converting `YearMonth` allows us to sort and plot data chronologically.  
> We also extract `year` and `month` for easier grouping later.

## Ensure Numeric Columns
These columns are numbers and need to be numeric for aggregation in SQL.

In [5]:
# Columns to convert to numeric (all except YearMonth)
numeric_cols = [
    'quantityPix', 'valuePix',
    'quantityTED', 'valueTED',
    'quantityTEC', 'valueTEC',
    'quantityBankCheck', 'valueBankCheck',
    'quantityBrazilianBoletoPayment', 'valueBrazilianBoletoPayment',
    'quantityDOC', 'valueDOC'
]

# Convert columns to numeric, coerce errors to NaN
df[numeric_cols] = df[numeric_cols].apply(pd.to_numeric, errors='coerce')

# Verify the conversion
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 15 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   YearMonth                       101 non-null    datetime64[ns]
 1   quantityPix                     101 non-null    float64       
 2   valuePix                        101 non-null    float64       
 3   quantityTED                     101 non-null    float64       
 4   valueTED                        101 non-null    float64       
 5   quantityTEC                     101 non-null    float64       
 6   valueTEC                        101 non-null    float64       
 7   quantityBankCheck               101 non-null    float64       
 8   valueBankCheck                  101 non-null    float64       
 9   quantityBrazilianBoletoPayment  101 non-null    float64       
 10  valueBrazilianBoletoPayment     101 non-null    float64       
 11  quanti

YearMonth                         0
quantityPix                       0
valuePix                          0
quantityTED                       0
valueTED                          0
quantityTEC                       0
valueTEC                          0
quantityBankCheck                 0
valueBankCheck                    0
quantityBrazilianBoletoPayment    0
valueBrazilianBoletoPayment       0
quantityDOC                       0
valueDOC                          0
year                              0
month                             0
dtype: int64

## Create Metrics: Average Transaction Values
For each payment method (PIX, TED, TEC, DOC, Boleto, Check), the average value per transaction is calculated as:

- **avg_method = valueMethod / quantityMethod**

In [6]:
# Calculate average transaction values for each payment method
df['avg_pix'] = df['valuePix'] / df['quantityPix']
df['avg_ted'] = df['valueTED'] / df['quantityTED']
df['avg_doc'] = df['valueDOC'] / df['quantityDOC']
df['avg_check'] = df['valueBankCheck'] / df['quantityBankCheck']
df['avg_boleto'] = df['valueBrazilianBoletoPayment'] / df['quantityBrazilianBoletoPayment']

df[['YearMonth','avg_pix','avg_ted','avg_doc','avg_check','avg_boleto']].head(20)

YearMonth   avg_pix    avg_ted   avg_doc  avg_check  avg_boleto
0  2024-05-01  0.408735  50.447408       NaN   3.833982    1.449973
1  2024-04-01  0.423123  51.702107       NaN   3.872846    1.450695
2  2024-03-01  0.394867  49.331041       NaN   3.620123    1.394168
3  2024-02-01  0.393443  45.303351  0.552755   3.574174    1.393646
4  2024-01-01  0.415980  47.443711  1.026333   3.491524    1.421268
5  2023-12-01  0.402509  45.257564  0.991070   3.487683    1.435014
6  2023-11-01  0.411429  42.151391  1.010217   3.432853    1.420785
7  2023-10-01  0.409384  46.266092  1.009969   3.615361    1.412479
8  2023-09-01  0.400089  45.960997  0.960279   3.495477    1.385345
9  2023-08-01  0.405954  44.491967  0.608641   3.581967    1.389036
10 2023-07-01  0.402682  46.884074  0.932032   3.413579    1.356361
11 2023-06-01  0.414267  49.689276  1.051294   3.438771    1.363460
12 2023-05-01  0.421116  49.593866  1.201194   3.597436    1.390904
13 2023-04-01  0.409408  46.683522  0.958716   3.560782    1.358958
14 2023-03-01  0.426497  45.146525  0.988495   4.068576    1.344983
15 2023-02-01  0.415749  41.140631  0.971865   4.110886    1.311795
16 2023-01-01  0.431494  43.530157  1.013847   3.313184    1.343002
17 2022-12-01  0.424599  41.706002  0.970754   3.336096    1.370083
18 2022-11-01  0.430174  39.999068  0.981210   3.349710    1.356501
19 2022-10-01  0.429456  40.685700  0.961264   3.306526    1.350127

> **Note:** Months with zero transactions (quantity = 0) will result in `NaN`.  
> This indicates that no transactions occurred that month, which is expected.

## Save Cleaned Dataset
Saving the cleaned dataset so it can be imported into SQL for aggregation and analysis:

In [7]:
df.to_csv('../data/cleaned/brazilian_payment_methods_clean.csv', index=False) # Save cleaned data to a new CSV file